libraries/Parsers/Microsoft.Bot.Builder.Parsers.LU/Parser/Visitor.cs

```diff
@@ -0,0 +1,207 @@
+            }
+
+            return new UtteranceAndEntitiesMap
+            {
+                Utterance = utterance.Trim(),
+                Entities = entities,
+                ErrorMsgs = errorMessages
+            };
+        }
+
+        private static string RecursivelyResolveTokenizedUtterance(List<object> tokenizedUterance, List<EntityElement> entities, List<string> errorMessages, string srcUtterance)
+        {
+            char[] invalidCharsInIntentOrEntityName = { '<', '>', '*', '%', '&', ':', '\\', '$' };
+            foreach (var item in tokenizedUterance)
+            {
+                if (item is AuxEntity)
+                {
+                    var auxEntity = (AuxEntity)item;
+                    var entityName = auxEntity.EntityName.Trim();
+                    if (!string.IsNullOrEmpty(entityName) && entityName.IndexOfAny(invalidCharsInIntentOrEntityName) >= 0)
+                    {
+                        errorMessages.Add($"Invalid utterance line, entity name {entityName} cannot contain any of the following characters: [<, >, *, %, &, :, \\, $]");
+                        continue;
+                    }
+
+                    if (auxEntity.EntityValue == null)
+                    {
+                        // we have a pattern.any entity
+                        var patternStr = !string.IsNullOrEmpty(auxEntity.Role) ? $"{{{auxEntity.EntityName}:{auxEntity.Role}}}" : $"{{{auxEntity.EntityName}}}";
+                        srcUtterance += patternStr;
+                        entities.Add(
+                            new EntityElement
+                            {
+                                Type = TypeEnum.PatternAnyEntities,
+                                Entity = auxEntity.EntityName.Trim(),
+                                Role = auxEntity.Role.Trim()
+                            });
+                    }
+                    else
+                    {
+                        // we have a new entity
+                        var newEntity = new EntityElement
+                        {
+                            Type = TypeEnum.Entities,
+                            Entity = auxEntity.EntityName.Trim(),
+                            Role = auxEntity.Role.Trim(),
+                            StartPos = srcUtterance.Length,
+                            EndPos = null
+                        };
+                        if (auxEntity.EntityValue == null)
+                        {
+                            errorMessages.Add($"Composite entity {auxEntity.Parent.EntityName} includes pattern.any entity {auxEntity.EntityName}.\r\n\tComposites cannot include pattern.any entity as a child.");
+                        }
+                        else
+                        {
+                            srcUtterance = RecursivelyResolveTokenizedUtterance(auxEntity.EntityValue, entities, errorMessages, srcUtterance).TrimStart();
+                            newEntity.EndPos = srcUtterance.Length - 1;
```

Comment:
Could some weird condition take this to negative or < startpos?

Reply:
I don't think so, I also took it into consideration but from my test from the JS version looks like not.